File: mobile/src/full/java/org/openhab/habdroid/ui/MapViewHelper.java
package org.openhab.habdroid.ui;

import android.app.AlertDialog;
import android.content.res.Resources;
import android.location.Location;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import org.openhab.habdroid.R;
import org.openhab.habdroid.core.connection.Connection;
import org.openhab.habdroid.model.Item;
import org.openhab.habdroid.model.ParsedState;
import org.openhab.habdroid.model.Widget;
import org.openhab.habdroid.util.Util;

import java.util.ArrayList;
import java.util.Locale;

public class MapViewHelper {
    public static WidgetAdapter.ViewHolder createViewHolder(LayoutInflater inflater,
            ViewGroup parent, Connection connection, WidgetAdapter.ColorMapper colorMapper) {
        MapsInitializer.initialize(inflater.getContext());
        return new GoogleMapsViewHolder(inflater, parent, connection, colorMapper);
    }

    private static class GoogleMapsViewHolder extends WidgetAdapter.LabeledItemBaseViewHolder
            implements GoogleMap.OnMarkerDragListener {
        private final MapView mMapView;
        private final int mRowHeightPixels;
        private GoogleMap mMap;
        private Item mBoundItem;
        private boolean mStarted;

        public GoogleMapsViewHolder(LayoutInflater inflater, ViewGroup parent,
                Connection connection, WidgetAdapter.ColorMapper colorMapper) {
            super(inflater, parent, R.layout.openhabwidgetlist_mapitem, connection, colorMapper);

            mMapView = itemView.findViewById(R.id.mapview);

            mMapView.onCreate(null);
            mMapView.getMapAsync(map -> {
                mMap = map;
                UiSettings settings = map.getUiSettings();
                settings.setAllGesturesEnabled(false);
                settings.setMapToolbarEnabled(false);
                map.setOnMarkerClickListener(marker -> { openPopup(); return true; });
                map.setOnMapClickListener(position -> openPopup());
                applyPositionAndLabel(map, 15.0f, false);
            });

            final Resources res = itemView.getContext().getResources();
            mRowHeightPixels = res.getDimensionPixelSize(R.dimen.row_height);
        }

        @Override
        public void bind(Widget widget) {
            super.bind(widget);

            ViewGroup.LayoutParams lp = mMapView.getLayoutParams();
            int rows = widget.getHeight() > 0 ? widget.getHeight() : 5;
            int desiredHeightPixels = rows * mRowHeightPixels;
            if (lp.height != desiredHeightPixels) {
                lp.height = desiredHeightPixels;
                mMapView.setLayoutParams(lp);
            }

            mBoundItem = widget.getItem();
            if (mMap != null) {
                mMap.clear();
                applyPositionAndLabel(mMap, 15.0f, false);
            }
        }

        @Override
        public void start() {
            super.start();
            if (!mStarted) {
                mMapView.onStart();
                mMapView.onResume();
                mStarted = true;
            }
        }

        @Override
        public void stop() {
            super.stop();
            if (mStarted) {
                mMapView.onPause();
                mMapView.onStop();
                mStarted = false;
            }
        }

        @Override
        public void onMarkerDragStart(Marker marker) {
            // no-op, we're interested in drag end only
        }

        @Override
        public void onMarkerDrag(Marker marker) {
            // no-op, we're interested in drag end only
        }

        @Override
        public void onMarkerDragEnd(Marker marker) {
            String newState = String.format(Locale.US, "%f,%f",
                    marker.getPosition().latitude, marker.getPosition().longitude);
            Item item = (Item) marker.getTag();
            Util.INSTANCE.sendItemCommand(mConnection.getAsyncHttpClient(), item, newState);
        }

        private void openPopup() {
            final MapView mapView = new MapView(itemView.getContext());
            mapView.onCreate(null);

            AlertDialog dialog = new AlertDialog.Builder(itemView.getContext())
                    .setView(mapView)
                    .setCancelable(true)
                    .setNegativeButton(R.string.close, null)
                    .create();

            dialog.setOnDismissListener(dialogInterface -> {
                mapView.onPause();
                mapView.onStop();
                mapView.onDestroy();
            });
            dialog.setCanceledOnTouchOutside(true);
            dialog.show();

            mapView.onStart();
            mapView.onResume();
            mapView.getMapAsync(map -> {
                map.setOnMarkerDragListener(GoogleMapsViewHolder.this);
                applyPositionAndLabel(map, 16.0f, true);
            });
        }

        private void applyPositionAndLabel(GoogleMap map, float zoomLevel, boolean allowDrag) {
            if (mBoundItem == null) {
                return;
            }
            boolean canDragMarker = allowDrag && !mBoundItem.getReadOnly();
            if (!mBoundItem.getMembers().isEmpty()) {
                ArrayList<LatLng> positions = new ArrayList<>();
                for (Item item : mBoundItem.getMembers()) {
                    LatLng position = toLatLng(item.getState());
                    if (position != null) {
                        setMarker(map, position, item, item.getLabel(), canDragMarker);
                        positions.add(position);
                    }
                }
                if (!positions.isEmpty()) {
                    LatLngBounds.Builder boundsBuilder = new LatLngBounds.Builder();
                    for (LatLng position : positions) {
                        boundsBuilder.include(position);
                    }
                    map.moveCamera(CameraUpdateFactory.newLatLngBounds(boundsBuilder.build(), 0));
                    float zoom = map.getCameraPosition().zoom;
                    if (zoom > zoomLevel) {
                        map.moveCamera(CameraUpdateFactory.zoomTo(zoomLevel));
                    }
                }
            } else {
                LatLng position = toLatLng(mBoundItem.getState());
                if (position != null) {
                    setMarker(map, position, mBoundItem, mLabelView.getText(), canDragMarker);
                    map.moveCamera(CameraUpdateFactory.newLatLngZoom(position, zoomLevel));
                }
            }
        }

        private static void setMarker(GoogleMap map, LatLng position, Item item,
                CharSequence label, boolean canDrag) {
            MarkerOptions marker = new MarkerOptions()
                    .draggable(canDrag)
                    .position(position)
                    .title(label != null ? label.toString() : null);
            map.addMarker(marker).setTag(item);
        }

        private static LatLng toLatLng(ParsedState state) {
            Location location = state != null ? state.getAsLocation() : null;
            return location != null
                    ? new LatLng(location.getLatitude(), location.getLongitude())
                    : null;
        }
    }
}

File: mobile/src/full/java/org/openhab/habdroid/ui/MapViewHelper.kt
/*
 * Copyright (c) 2010-2024 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.ui

import android.location.Location
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.view.isGone
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.MapsInitializer
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import java.util.Locale
import org.openhab.habdroid.R
import org.openhab.habdroid.model.Item
import org.openhab.habdroid.model.Widget

object MapViewHelper {
    fun createViewHolder(initData: WidgetAdapter.ViewHolderInitData): WidgetAdapter.ViewHolder {
        MapsInitializer.initialize(initData.inflater.context)
        return GoogleMapsViewHolder(initData)
    }

    private class GoogleMapsViewHolder(initData: WidgetAdapter.ViewHolderInitData) :
        WidgetAdapter.AbstractMapViewHolder(initData) {
        private val mapView = baseMapView as MapView
        private var map: GoogleMap? = null

        init {
            mapView.onCreate(null)
        }

        override fun bindAfterDataSaverCheck(widget: Widget) {
            super.bindAfterDataSaverCheck(widget)
            withLoadedMap { map ->
                map.clear()
                map.applyPositionAndLabel(widget, 15.0f, false)
            }
        }

        override fun onStart() {
            super.onStart()
            mapView.onStart()
            mapView.onResume()
        }

        override fun onStop() {
            super.onStop()
            mapView.onPause()
            mapView.onStop()
        }

        override fun openPopup() {
            val widget = boundWidget ?: return
            fragmentPresenter.showBottomSheet(MapBottomSheet(), widget)
        }

        private fun withLoadedMap(callback: (map: GoogleMap) -> Unit) {
            val loadedMap = this.map
            if (loadedMap != null) {
                callback(loadedMap)
                return
            }

            mapView.getMapAsync { map ->
                this.map = map
                with(map.uiSettings) {
                    setAllGesturesEnabled(false)
                    isMapToolbarEnabled = false
                }
                map.setOnMarkerClickListener {
                    openPopup()
                    true
                }
                map.setOnMapClickListener { openPopup() }
                callback(map)
            }
        }
    }
}

fun GoogleMap.applyPositionAndLabel(widget: Widget, zoomLevel: Float, allowDrag: Boolean) {
    val item = widget.item ?: return
    val canDragMarker = allowDrag && !item.readOnly
    if (item.members.isNotEmpty()) {
        val positionMap = item.members
            .map { m -> m.state?.asLocation?.toLatLng()?.let { m to it } }
            .filterNotNull()
            .toMap()
        if (positionMap.isNotEmpty()) {
            val boundsBuilder = LatLngBounds.Builder()
            positionMap.forEach { member, pos ->
                setMarker(pos, member, member.label, canDragMarker)
                boundsBuilder.include(pos)
            }
            moveCamera(CameraUpdateFactory.newLatLngBounds(boundsBuilder.build(), 0))
            if (cameraPosition.zoom > zoomLevel) {
                moveCamera(CameraUpdateFactory.zoomTo(zoomLevel))
            }
        }
    } else {
        item.state?.asLocation?.toLatLng()?.let { pos ->
            setMarker(pos, item, widget.label, canDragMarker)
            moveCamera(CameraUpdateFactory.newLatLngZoom(pos, zoomLevel))
        }
    }
}

fun GoogleMap.setMarker(position: LatLng, item: Item, label: CharSequence?, canDrag: Boolean) {
    val marker = MarkerOptions()
        .draggable(canDrag)
        .position(position)
        .title(label?.toString())
    addMarker(marker)?.tag = item
}

fun Location.toLatLng() = LatLng(latitude, longitude)

fun Location.toMapsUrl() = "https://www.google.de/maps/@$latitude,$longitude,16z"

class MapBottomSheet :
    AbstractWidgetBottomSheet(),
    GoogleMap.OnMarkerDragListener {
    private lateinit var mapView: MapView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.bottom_sheet_map, container, false)
        val title = view.findViewById<TextView>(R.id.title)

        title.text = widget.label
        title.isGone = widget.label.isEmpty()

        mapView = view.findViewById(R.id.mapview)
        mapView.onCreate(null)

        return view
    }

    override fun onDestroyView() {
        mapView.onDestroy()
        super.onDestroyView()
    }

    override fun onStart() {
        super.onStart()
        mapView.onStart()
    }

    override fun onStop() {
        super.onStop()
        mapView.onStop()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()

        mapView.getMapAsync { map ->
            map.setOnMarkerDragListener(this@MapBottomSheet)
            map.applyPositionAndLabel(widget, 16.0f, true)
        }
    }

    override fun onPause() {
        mapView.onPause()
        super.onPause()
    }

    override fun onMarkerDragStart(marker: Marker) {
        // no-op, we're interested in drag end only
    }

    override fun onMarkerDrag(marker: Marker) {
        // no-op, we're interested in drag end only
    }

    override fun onMarkerDragEnd(marker: Marker) {
        val newState = String.format(Locale.US, "%f,%f", marker.position.latitude, marker.position.longitude)
        connection?.httpClient?.sendItemCommand(marker.tag as Item?, newState)
    }
}
